### PR TITLE
Add `MockBackend`

### PR DESCRIPTION
**Pull request by georgwiese**
Adds a basic mock prover that simply asserts that all constraints are satisfied.

This first version has the following features:
- Polynomial constraints are verified.
- Speed is reasonable. For example, the RISC-V Keccak example is verified in 0.087s.
- Error messages are good: It prints the relevant constraint and row, and all relevant assignments.

Missing features:
- Machine connections (i.e., any lookups or permutations) are not yet validated.
- Later-stage witnesses are not yet validated.

Caveats:
- We rely on `powdr_backend_utils::split_pil`, so lookups / permutations *within* a namespace are ignored.

Example:
```
$ cargo run pil test_data/pil/fibonacci.pil -o output -f --prove-with mock
```

### CODE REVIEW

**Comment by lvella** [comment]
doesn't draft implies [WIP] ?

**Comment by georgwiese** [comment]
> doesn't draft implies [WIP] ?

It's me trying not to get any comments on a PR, but obviously it's not working 😛 

I usually put `[WIP]` in the title to make it more clear visually. Can also stop doing that, but I'm not sure everyone sees the draft status. I don't want someone to waste time reviewing a PR that's not ready.

**Review by leonardoalt — COMMENTED** [review]
Nice!

File: backend/src/lib.rs [hunk 1]
```diff
@@ -69,6 +72,7 @@ pub const DEFAULT_ESTARK_OPTIONS: &str = "stark_gl";
 impl BackendType {
     pub fn factory<T: FieldElement>(&self) -> Box<dyn BackendFactory<T>> {
         match self {
+            BackendType::Mock => Box::new(mock::MockBackendFactory::new()),
```
[the remark this answers is not in the document]
Comment (georgwiese) [reply backend/src/lib.rs:75]: Done!

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**